src/lib/structures/modelMap.ml

```diff
@@ -127,12 +127,45 @@ let add ((id, arg_tys, _) as sy) arg_vals ret_val { values; suspicious } =
   if List.compare_lengths arg_tys arg_vals <> 0 then
     Fmt.invalid_arg "The arity of the symbol %a doesn't agree the number of \
                      arguments" Id.pp id;
-
-  let graph = try P.find sy values with Not_found -> Graph.empty in
+  let graph =
+    try
+      let graph = P.find sy values in
+      (* If the graph associated with [sy] contains only an abstract value,
+         it means there is no constraint on this graph. We replace it by
+         the graph with the only constraint given by [arg_vals] and
+         [ret_val]. *)
+      if Graph.cardinal graph == 1 then
+        let _, value = Graph.choose graph in
+        let Expr.{ f; _ } = Expr.term_view value in
+        match f with
+        | Sy.Name { hs; _ }
+          when Id.Namespace.Abstract.is_id (Hstring.view hs) ->
+          Graph.empty
+        | _ -> graph
+      else
+        graph
+    with Not_found -> Graph.empty in
   let values = P.add sy (Graph.add arg_vals ret_val graph) values in
   { values; suspicious }
 
-let empty ~suspicious = { values = P.empty; suspicious }
+module DeclSets = Set.Make
+    (struct
+      type t = Id.typed
+      let compare = Id.compare_typed
+    end)
+
+let empty ~suspicious declared_ids =
+  let values =
+    List.fold_left
+      (fun values ((_, arg_tys, ret_ty) as sy) ->
+         let arg_vals = List.map Expr.mk_abstract arg_tys in
+         let ret_val = Expr.mk_abstract ret_ty in
+         let graph = Graph.add arg_vals ret_val Graph.empty in
+         P.add sy graph values
```

Comment:
It looks like you are using a graph with a single "dummy" definition here to then detect it in `add` and fix things up if the graph actually has definitions. This is quite hacky in part because it introduces a global invariant that abstract values are never otherwise placed into graphs — I believe it is true currently (although I actually had to check in several places to say so) but could easily cause subtle bugs in the future due to unrelated refactorings.

Your suggestion in the PR message to use an ADT (or a record carrying an explicit default value) would be better here; I don't think the extra indirection will be a problem (especially given that it will only impact model generation).

Reply:
I know this is a hackish solution. I think the best solution would be to add a function to generate dummy values. If you run `cvc5` or `z3` on an input file with completely unused symbols, they generate dummy values for them. I didn't do it in this PR because I want a small here in order to implement `(get-value)`.

Besides, I'm not sure to understand what you mean by a global invariant here. The `add` replaces a valid model by another valid model and you don't need to maintain an invariant outside `ModelMap` because the only way to add new constraints in your graph is by using `ModelMap.add`.

I agree the indirection isn't a big deal.

Reply:
> I'm not sure to understand what you mean by a global invariant here. The add replaces a valid model by another valid model and you don't need to maintain an invariant outside ModelMap because the only way to add new constraints in your graph is by using ModelMap.add.

I mean that in the current implementation it is incorrect to call `ModelMap.add` with a `ret_val` that is an abstract value:

 - The graph is initially `{ [@a0] -> @a1 }` as created by `empty`
 - Adding the mapping `0 -> @x`, the graph becomes `{ [0] -> @x }` because it only contains a single abstract value
 - Adding the mapping `1 -> @y`, the graph becomes `{ [1] -> @y }` because it only contains a single abstract value

But this is incorrect and now we will print `(define-fun f ((_ Int)) T @y)` but `(f 0)` must be `@x`, not `@y` (abstract values with different names are distinct).

Reply:
Uhm right, it should be fine with the current state of the code but it's a fragile design. I'll fix it.